### PR TITLE
Improved mountain tool cost

### LINKED ISSUE #2127: Multi-tile mountain landscaping prices
Raising a mountain with a one-tile-square selection costs $80.  Considering that nine tiles are affected, one might think the cost should be closer to $20*9 = $180, but Chris Sawyer's landscaping service offers a discount.  (If I recall correctly, this is consistent with RCT1.)

1 tile mountain

With OpenRCT2, we can build mountains that have larger plateaus.  A mountain with a top two tiles square costs $320 to form.  Seeing as 16 tiles are affected, $20 \* 16 = $320 sounds like it's about the right price to pay.

2 tile mountain

By the time we reach a 3x3 mountain, the cost is $720.  25 tiles are affected, so we'd expect to pay closer to $20*25 = $500.  That said, some of the tiles would have taken two $20 operations to form, so it's about right regardless.

3 tile mountain

A 4x4 mountain costs $1,280 to start, which is well above the $20 \* 36 = $720 we might expect to pay for it.  This is the turning point:  we could have built the same shape for less money using the standard tools.

4 tile mountain

It doesn't get better from here.  Starting a 64x64 mountain costs $327,680, which is way above the estimated $87,120.

It seems that the mountain tool is very naive in its pricing:  it simply multiplies the one-tile price by the number of tiles in the selection.  This isn't really the best way to go about it because it overcounts redundant land-moving operations.  I propose we change the pricing scheme to evaluate how much money it would cost if tiles were moved manually and use that value instead of the naive multiplication.

### LINKED ISSUE #2229: Mountain tool only charges for scenery removed at cursor
Yes, more issues with the mountain tool...

If using the mountain tool causes scenery to be removed, you'll only be charged/refunded for the scenery at the mountain tool's cursor.  I'm not sure if this is consistent with RCT1 or not.  I tested this in OpenRCT2 build 1421.

trees
trees_after_mountain
trees_2
trees_after_mountain_2

### PR DESCRIPTION
**Pull request by jensj12**
Fix #2127, #2229, #5586
Replaces #5332
The cost is now based on the number of quarter cells changed. This applies to the normal landscaping tool as well. The landscaping tool charges $5 per quarter tile. Raising a full tile still costs $20 without and $80 with the mountain tool.
Requires network version increase.

### CODE REVIEW

File: src/openrct2/world/map.c [hunk 1]
```diff
@@ -1747,6 +1792,12 @@ static money32 map_set_land_height(sint32 flags, sint32 x, sint32 y, sint32 heig
         } while (!map_element_is_last_for_tile(mapElement++));
     }
 
+	for (int i = 0; i < 4; i += 1) {
```
Comment (HaasJona) [inline src/openrct2/world/map.c]: The rest of the file uses spaces for indentation, you use tabs here

File: src/openrct2/world/map.c [hunk 2]
```diff
@@ -1580,6 +1580,52 @@ static sint32 map_set_land_height_clear_func(rct_map_element** map_element, sint
     return 1;
 }
 
+static sint32 map_get_corner_height(sint32 z, sint32 slope, sint32 direction)
+{
+	switch (direction) {
```
Comment (HaasJona) [inline src/openrct2/world/map.c]: Tab indentation

**Comment by jensj12** [comment]
Rebased and ready to go.
There are still a small bug present that I will fix in another PR, as it requires changing the algorithm that smooths the land. Along the sides of the selected area, it may think it is going to remove a piece of scenery twice or (when raising a sloped tile at the highest corner) raise a corner of land twice, but unless the removed scenery gains money, the correct amount is charged.

**Comment by Broxzier** [comment]
Tested it, and the costs are much more as expected. One note though, my frame rate drops from around 50 fps to only 6 fps when the mountain tool is active (debug build). I assume that is because `map_get_lowest_land_height` and `map_get_highest_land_height` get called in the update function. Is it possible to move these calls to only be used when the `GAME_COMMAND_FLAG_APPLY` is set?

**Comment by jensj12** [comment]
Those calls are essential for the cost calculation. If I don't know what the selected land is going to look like, I can't predict the cost. Only changing the land when `GAME_COMMAND_FLAG_APPLY` is set is basically what the old code did.

Do you also experience the lag without this patch? The mountain tool currently always tests the entire map. Perhaps we need to cache the result in the future.

**Comment by Broxzier** [comment]
> Do you also experience the lag without this patch?

Yes, though slightly less. 50 fps -> 11 fps. Caching and checking if the land tool cursor has been moved beforehand would probably work, but I guess that better done in a separate PR.

**Comment by jensj12** [comment]
I'm making some good progress on a rewrite on the mountain tool algorithm (which fixes the fps drop and the minor pricing issues left), but it depends on this PR. Since it solves different issues and requires more testing, I'd rather make a new PR for it.
Can I get a notification when this is ready to be merged so I can rebase?

**Pull request merged**